march_hardware_interface/src/march_hardware_interface.cpp

```diff
@@ -316,4 +313,25 @@ void MarchHardwareInterface::updatePowerNet()
     }
   }
 }
+
+void MarchHardwareInterface::outsideLimitsCheck(int joint_index)
+{
+  march4cpp::Joint joint = marchRobot.getJoint(joint_names_[joint_index]);
+  if (joint_position_[joint_index] < soft_limits_[joint_index].min_position ||
+      joint_position_[joint_index] > soft_limits_[joint_index].max_position)
+  {
+    ROS_ERROR_THROTTLE(1, "Joint %s is outside of its soft_limits_ (%f, %f). Actual position: %f",
+                       joint_names_[joint_index].c_str(), soft_limits_[joint_index].min_position,
+                       soft_limits_[joint_index].max_position, joint_position_[joint_index]);
+
+    if (joint.canActuate())
+    {
+      std::ostringstream errorStream;
+      errorStream << "Joint " << joint_names_[joint_index].c_str() << " is out of its soft_limits_ ("
+                  << soft_limits_[joint_index].min_position << ", " << soft_limits_[joint_index].max_position
+                  << "). Actual position: " << joint_position_[joint_index];
+      throw ::std::invalid_argument(errorStream.str());
```

Comment:
I do not entirely understand what is happening if the joint is outside of it's softlimits, while able to actuate. It throws an error, but what is happening then? Will all the motors shut down? Does the entire system shut down?

What I have found:
Error: Definitely a problem that should be investigated. SysAdmin should be notified automatically, but doesn't need to be dragged out of bed.
Fatal/Critical: Overall application or system failure that should be investigated immediately. Yes, wake up the SysAdmin. 

I think that we have to investigate this kind of fault immediately. But you probably have a reasoning behind it. Curious to hear it.

Reply:
When an error is thrown (which is not catched/handled) the node will crash, this means in this case:
- The hardware_interface crashes
- SOEM communication stops
- All IMC's goes to fault state
(If I'm correct)

Reply:
I changed the invalid argument to run-time error. This is more appropriate. We are still discussing how to handle hardware_interface crashes. For now the safety node is not responding to this, but in the future we probably want to add a ROS_FATAL in the safety node when the hardware_interface crashes.